# Video Classification

Video classification is one capability among many under the broad umbrella of _video understanding_, technologies that automatically extract information from video. You can read more about the great, wide world of video understanding in our blog post [here](https://www.notion.so/fastforwardlabs/December-2021-ccbd4d6633dc40d0b329c41f71bb6406#f6e2430cacdb434ea7fef5d4a1535ce4). Video classification is one of the fundamental building blocks used in a slew of video understanding tasks. Its goal is to identify the action or motion within a video. This capability is similar to image clasification but with a twist -- models must not only detect the cat, but also what the cat is _doing_. 

The goal of this notebook is to provide an introduction to video classification, including datasets, models, and methods used for video classification and action recognition. Specifically, we'll explore the Kinetics video dataset and the pre-trained I3D model for action recognition, hosted on the Tensorflow Model Hub. This notebook requires Tensorflow as well as a few other packages, which we install below. 

## Setting up

In [ ]:
%%capture 
#hides cell output. disable for logs.
!pip install -r requirements.txt

In [2]:
from collections import defaultdict
import os
import glob
import numpy as np
import pandas as pd
import tensorflow as tf
from IPython import display

In order to make it easier to work with the dataset and model in this notebook, we created a small library of helper functions and classes. Here, we import the good bits. 

In [3]:
from vidbench.data.load import KineticsLoader
from vidbench.models import I3DLoader
from vidbench.predict import predict, store_results_in_dataframe, compute_accuracy, evaluate
from vidbench.visualize import make_video_table
from vidbench.data.process import resample_video, load_and_resize_video, video_acceptable

While this notebook does not require a GPU to run (we only process a few videos for demonstration purposes), if you have GPUs at your disposal you can check them here. 

In [ ]:
print("Num GPUs Available: ", len(tf.config.list_physical_devices('GPU')))

## The Data

In this notebook we make use of the DeepMind [Kinetics dataset](https://arxiv.org/abs/1705.06950), which consists of thousands of YouTube videos focused on human actions and interactions. While there are several versions of this dataset, we'll focus primarily on the Kinetics 400 dataset which contains 400 human action classes, ranging from human-object interactions like playing instruments, as well as human-human interactions like shaking hands. Each video clip is approximately ten seconds and has been sourced from a unique YouTube video.   


As of November 2021, the video files that make up the Kinetics datasets are stored at https://s3.amazonaws.com/kinetics/ as described in this [repository](https://github.com/cvdfoundation/kinetics-dataset). The Kinetics 400 contains a total of 306,245 video clips with at least 400 clips in each class.  These are distributed among three splits: training, test and validation. Each split consists of thousands of videos in `.mp4` format. 

| Dataset split | Clips per class | Total Videos |
|---------------|-----------------|--------------|
| train         |  250-1000       |    246245    |
| test          |   100           |   40000      |
| val           |     50          |   20000      |


For each dataset split, videos are grouped into a series of directories and each directory is packaged as a `tar.gz` file that needs to be unpacked. Each of these `tar.gz` files contains about 1000 video clips.  In this notebook we'll explore a handful of videos from the validation set and we created a `KineticsLoader` class to handle downloading and unpacking these video files.  While this class is designed to handle the full validation (or test, or train) set, we can also use it to explore a small portion of the videos, which we'll demonstrate in the following cells. 

In [4]:
# this class handles the infrastructure to support downloading, unpacking, and pre-processing videos
loader = KineticsLoader(version="400", split="val")

No need to fetch, path already exists /home/cdsw/data/raw/kinetics/400/val/k400_val_path.txt
No need to fetch, path already exists /home/cdsw/data/raw/kinetics/400/val/val.csv


If you're exploring this notebook after performing automatic setup through the CML AMP interface, then we've already downloaded a chunk of videos to explore. If not, running the cell below will initiate a download and unpacking of (at least) 500 vidoes (recall that they are grouped in chunks of approximately 1000). 

In [ ]:
loader.download_n_videos(500)

With the code above we've downloaded just 1000 videos from the validation set (out of 20,000 available videos). We could download the entire validation set but that would require at least 30-50 GBs of storage! Since our goal in this notebook is simply to explore the video classification capability, 1000 videos is more than enough to explore with. This AMP also includes a benchmarking script that will allow the user to evaluate a model on _all_ videos in any of the data splits discussed above (more on this towards the end of the notebook).   

We've also downloaded the ground truth labels for _all_ 20K videos in the validation set. These are stored in a Pandas DataFrame which you can see below.

In [6]:
ground_truth_labels_df = pd.read_csv(f"{loader.data_dir}/val.csv")
ground_truth_labels_df

label   youtube_id  time_start  time_end split  is_cc
0      abseiling  0wR5jVB-WPk         417       427   val      0
1      abseiling  3caPS4FHFF8          36        46   val      0
2      abseiling  3yaoNwz99xM          62        72   val      1
3      abseiling  6IbvOJxXnOo          47        57   val      0
4      abseiling  6_4kjPiQr7w         191       201   val      0
...          ...          ...         ...       ...   ...    ...
19901      zumba  w5hbJLVhZDI          93       103   val      0
19902      zumba  xDd6uIBeMEA           1        11   val      0
19903      zumba  XWvGn7eI04A          12        22   val      0
19904      zumba  yGdQwxP5koA          83        93   val      0
19905      zumba  ZVDR2od1gn8          37        47   val      0

[19906 rows x 6 columns]

As you can see, this DataFrame has nearly 20K rows - one for each video clip in the validation set. We need to filter this DataFrame to just those videos that we've actually downloaded. Our `KineticsLoader` keeps track of which videos we've downloaded. We can see all available pathnames with the following line. We're only showing the top 20 but there are 1000 video pathnames available. 

In [11]:
loader.video_filenames[:20]

['/home/cdsw/data/raw/kinetics/400/val/part_0/0tlGOxUQ0Kw_000074_000084.mp4',
 '/home/cdsw/data/raw/kinetics/400/val/part_0/-05qSkAhM6Y_000205_000215.mp4',
 '/home/cdsw/data/raw/kinetics/400/val/part_0/-LISB_b8rIw_000049_000059.mp4',
 '/home/cdsw/data/raw/kinetics/400/val/part_0/--ILYNHl3e4_000541_000551.mp4',
 '/home/cdsw/data/raw/kinetics/400/val/part_0/-yv8c2CDbR8_000004_000014.mp4',
 '/home/cdsw/data/raw/kinetics/400/val/part_0/-aeOuOI3eN0_000219_000229.mp4',
 '/home/cdsw/data/raw/kinetics/400/val/part_0/-WZgMWx8Elk_000013_000023.mp4',
 '/home/cdsw/data/raw/kinetics/400/val/part_0/-Y-fUYGcb7o_000049_000059.mp4',
 '/home/cdsw/data/raw/kinetics/400/val/part_0/-beyXnxwTao_000040_000050.mp4',
 '/home/cdsw/data/raw/kinetics/400/val/part_0/-whdHn9Mbcc_000000_000010.mp4',
 '/home/cdsw/data/raw/kinetics/400/val/part_0/-CAPalSW0QI_000546_000556.mp4',
 '/home/cdsw/data/raw/kinetics/400/val/part_0/0HydDezkSU0_000018_000028.mp4',
 '/home/cdsw/data/raw/kinetics/400/val/part_0/-4hx9N2OhZo_000029

Next, we'll filter the ground truth DataFrame to include only those vidoes that have already been downloaded and are available locally. 

In [7]:
available_youtube_ids = []
for filename in loader.video_filenames:
    pathname, vidname = os.path.split(filename)
    available_youtube_ids.append(vidname[:11]) #youtube IDs are 11 characters long

available_videos_metadata_df = ground_truth_labels_df[ground_truth_labels_df['youtube_id'].isin(available_youtube_ids)]

In [6]:
available_videos_metadata_df

label   youtube_id  time_start  time_end split  is_cc
0                abseiling  0wR5jVB-WPk         417       427   val      0
50            air drumming  --nQbRBEz2s         104       114   val      0
99     answering questions  -egPJubR-CE           1        11   val      0
100    answering questions  -ejLPB4J4SM         106       116   val      0
101    answering questions  -emx5qjikEc           1        11   val      0
...                    ...          ...         ...       ...   ...    ...
19811                 yoga  -Gb1pbOE32g          32        42   val      0
19812                 yoga  -IeJ0CF3huY          16        26   val      0
19813                 yoga  -kSK2kqnTHA          22        32   val      0
19814                 yoga  0wHOYxjRmlw          41        51   val      1
19860                zumba  0H5mnFcm2Kg          17        27   val      0

[1000 rows x 6 columns]

Much better! Now we have a DataFrame with only 1000 rows -- one for each locally available video clip. What did we end up downloading? Let's take a look at what classes we have. 

In [15]:
available_videos_metadata_df.label.value_counts()

massaging legs              10
scrambling eggs              9
massaging person's head      8
eating chips                 8
folding paper                7
                            ..
lunge                        1
massaging feet               1
punching person (boxing)     1
blowing out candles          1
pumping fist                 1
Name: label, Length: 360, dtype: int64

Out of 400 unique classes, we have videos from 360 of those. Most classes only have one representative video clip, though we do have a handful of video clips from classes like "massaging legs," and "scrambling eggs." 

Video classification is computationally expensive so in the following cell we take a small, random sample of vidoes to explore for the remainder of this notebook. Because this is a random sample, each time you run this cell you'll get a new set of 8 videos to play with!

In [8]:
NUM_VIDEOS = 8

video_sample = available_videos_metadata_df.sample(NUM_VIDEOS)
video_sample

label   youtube_id  time_start  time_end split  is_cc
10302             parkour  -6KNHzXeYKo         173       183   val      0
1796        brushing hair  -Gio3hF5OA0          36        46   val      0
5923       exercising arm  0wZpjStZtUY           1        11   val      0
5471          eating cake  0pQACKjlllc           2        12   val      0
1599        braiding hair  0uswr636GWs          57        67   val      0
17866            tickling  -JuFFz7KyMo          41        51   val      0
11499       playing flute  -SdDjLIkBHc          43        53   val      0
6018   extinguishing fire  -X1Allr_ZcY          15        25   val      0

There we go! We've selected a manageable batch of just 8 videos to examine. And we can see at a glance which classes our model will be attempting to predict. 

###  Visualize some videos

So what kind of videos are we dealing with?  In the cell below we display our video clip sample along with their ground truth class labels. As you play each video, notice that each is only approximately ten seconds long. 

In [9]:
video_html = make_video_table(video_sample['youtube_id'].values, video_sample['label'].values)
display.HTML(video_html)

### Pre-processing

Now that we have a sense of what kind of videos we're working with, let's start classifying them! But before we do that, we have another step to perform -- preprocessing.  

The YouTube videos in the Kinetics dataset are all in `.mp4` format but TensorFlow models do not recognize this! We must convert the videos into a format that our TF model can work with. This requires two steps: 
1. convert the `.mp4` format to a more appropriate data structure, like NumPy arrays
2. Resize the video dimensions to work within model specifications

####  Video resizing 
While the first step is likely self-explanatory, the second step deserves some attention. Those with experience working with pre-trained image classification models are likely already familiar with the idea that these models require image inputs of a specific height and width. These requirements are determined during model training and set limits on how large (or small) an image must be in pixels in order for the model to process that image. Video classification is no different in this respect, but comes with a third dimension of complexity: time. 

Let's take a look at the dimensions of the videos in our sample batch. The following cells will read an `.mp4` video clip into a numpy array and print the shape of that array to the screen. The shape tuple has the following format: 

(number of frames, height in pixels, width in pixels, number of color channels)

In [10]:
def load_video(path):
    """Convert video to Numpy array."""
    import cv2
    cap = cv2.VideoCapture(path)
    frames = []
    try:
        while True:
            ret, frame = cap.read()  # frame is in BGR format
            if not ret:
                break
            frames.append(frame)
    finally:
        cap.release()
    return np.array(frames).astype("float32") / 255.0

In [11]:
video_paths = [glob.glob(f"{pathname}/{yt_id}*")[0] for yt_id in video_sample['youtube_id'].values]

for video_path in video_paths: 
    video_np = load_video(video_path)
    print(video_np.shape)

(150, 240, 320, 3)
(300, 360, 480, 3)
(234, 320, 568, 3)
(258, 360, 202, 3)
(300, 360, 480, 3)
(300, 320, 568, 3)
(300, 360, 486, 3)
(300, 720, 1280, 3)


As we can see, there's quite a bit of variation among our video batch that isn't really detectable when we viewed the raw video clips earlier. While each video is exactly 10 seconds long, some have 300 frames and others have only 127 frames. Some have small spatial dimensions (240 x 320) while others are quite large (720 x 1280). The only thing common to all videos is that they each have three color channels, the familiar RGB system. 

This has implications for how we sample and process these video clips for model consumption. The model we'll use in this notebook requires spatial dimensions of (224 x 224) so we'll need to resize and crop each frame of each video to these dimensions. 

But how do we deal with the temporal dimension, i.e., the number of frames?  That depends on how we want to use our video classification model. The model has no specified limit on the number of frames it can accept (note, however, that more frames translates to longer processing time which can become very computationally expensive!) If we only send to the model one video clip at a time, we can feed it the full number of frames for inference. However, it's usually faster to send a batch to the model rather than sending each video separately. In that case, we need to deal with the variation in frame rates for these videos because we cannot create a numpy array in which one of the dimesions varies!  

#### Video Resampling
The crux of the issue lies in the fact that different cameras have different frame capture rates. Higher quality videos have more frames-per-second (FPS) than lower quality videos leading to a situation in which a collection of 10 second videos can nevertheles have different numbers of frames. Below we show a toy example of two videos that are each 10 seconds long but the top fewer frames than the bottom one over that time span due to having a lower FPS.   

![Upsampling](images/video_fps_example.jpg)

In order to create a batch of video clips, we must resample the clips so that each has the same number of frames.  This can be accomplished either by _upsampling_ videos with low FPS, or _downsampling_ videos with high FPS.  A simple way to perform upsampling is to duplicate certain frames throughout the length of the video. In the figure below, we see Video 1 has frames added to match the number of frames contained in Video 2.  These frames are duplicates of existing frames in Video 1 (which is why some of the frames are repeated colors). 

<img src="images/upsample.gif">

In contrast, downsampling involves removing frames periodically throughout the video. This time, we remove a sample of unique frames from Video 2 so that it has the same number of frames as Video 1. 
<img src = "images/downsample.gif">

In either case, resampling should be done in such a way that the frames we duplicate or remove are equally dispersed throughout the duration of the 10 second clip in order to capture as much of the original motion as possible. 

In a pinch we could simply grab a fixed-size chunk of consecutive frames (from the beginning, middle, or end) from each video clip. The problem here is that this chunk may only represent a portion of the full 10 second interval. The model would then attempt to infer on a batch of videos in which one of them has frames representing close to the full 10 seconds, while another has frames representing only a fraction of the time. This could increase the difficulty of the model to properly classify the latter video.  

#### Load videos into NumPy arrays

Luckily, our `KineticsLoader` class has a `load_videos` method, similar to the one above, that also handles these pre-processing steps. Here's what we do under the hood:

1. First, a central square proportional to the size of the video is cropped
2. The cropped portion is resized to 224x224 pixels
3. Videos are resampled 
   - If the user provides `num_frames`, all videos are resampled to have this many frames (upsampling those that have lower FPS, downsampling those with higher FPS)
   - If `num_frames` is not provided, the algorithm determines which video in the batch has the lowest FPS (fewest frames) and all vidoes are downsampled to match this frame rate
   

Again we note that video clips with more total frames will take longer to process through the model. For the current dataset, `num_frames = 128` is reasonable choice.  For reference, the maximum number of frames that we've seen in the dataset is 300 (30 FPS). 

In [16]:
def load_videos(youtube_ids, num_frames):
    """Create numpy array with batch of videos from list of youtube ids""",
    # we created this list of video pathnames in an earlier cell
    global video_paths
    
    video_batch = []
    for video_path in video_paths:
        video_np = load_and_resize_video(video_path, resize_type="crop")
        video_np = np.expand_dims(video_np, axis=0)
        video_batch.append(video_np)
            
    video_batch_resampled = []
    for video in video_batch:
        resampled_video = resample_video(video, num_frames)
        video_batch_resampled.append(resampled_video)
            
    return np.concatenate(video_batch_resampled, axis=0).astype("float32")

In [17]:
num_frames = 128   
videos_tensor = load_videos(video_sample['youtube_id'].values, num_frames=num_frames)

In [18]:
videos_tensor.shape

(8, 128, 224, 224, 3)

We now have a batch of 8 video clips in a format that our model will understand. 

It's time to classify!

## The Model

In this notebook we make use of the Inflated 3D ConvNet (I3D) video classification model. This model architecture was introduced in 2017 and provided state-of-the-art results for video action classification for multiple datasets. We dig into some of the details about the architecture and training for this model in our accompanying blog post [TODO: Link in January]. You can also read more about this model in the original paper, [Quo Vadis, Action Recognition? A New Model and the Kinetics Dataset](https://arxiv.org/abs/1705.07750). 

Since it's inception, there are now multiple pre-trained versions of the I3D model that are publicly available on the [TensorFlow Model Hub](https://www.tensorflow.org/hub). The original version was pre-trained on the Kinetics 400 dataset, which we explored above. Another version was trained on the Kinetics 600 dataset. We created an `I3DLoader` class to handle model loading from the TF Model Hub. You can choose either the `kinetics-400` or `kinetics-600` version. 

The biggest difference between these two models is in the number of classes they predict. I3D trained on Kinetics 400 predicts, you guessed it, 400 different classes, while it's counterpart (I3D trained on Kinetics 600) predicts on 600 classes. While we didn't discuss the Kinetics 600 dataset in detail, it is essentially a superset of the Kinetics 400 dataset -- it includes all 400 labels from that dataset, plus an additional 200 unique classes. Either can be used to make inference on our sample of videos but keep in mind that with more classes comes more challenges in predicting the correct class -- there are simply more options for the model to choose from. 

Below we load up the I3D model trained on the Kinetics 400 dataset. 

In [19]:
i3d400 = I3DLoader(trained_on='kinetics-400')

Let's look at what kinds of predictions this model can make. 

In [13]:
i3d400.labels[:20]

['abseiling',
 'air drumming',
 'answering questions',
 'applauding',
 'applying cream',
 'archery',
 'arm wrestling',
 'arranging flowers',
 'assembling computer',
 'auctioning',
 'baby waking up',
 'baking cookies',
 'balloon blowing',
 'bandaging',
 'barbequing',
 'bartending',
 'beatboxing',
 'bee keeping',
 'belly dancing',
 'bench pressing']

### Get model predictions

Now that we have data in the proper format, ground truth labels, and a model loaded and ready to go -- it's time to make predictions! This notebook was developed on CPUs so the following cell can take some time to run. One way to speed up inference is to resample the vidoes to each have fewer frames, as we discussed above. The downside is that performance will likely degrade since the model will have fewer frames on which to make a prediction.

In [20]:
scores, predictions, _, _ = predict(videos_tensor, i3d400, verbose=False)

The output of our `predict` function includes the probabilities associated with each of the 400 labels and the 400 labels themselves for each video clip in our batch, sorted in descending order so that the most likely classes (highest probabilities) are at the top of the list. 

In [21]:
predictions

array([['parkour', 'hopscotch', 'skateboarding', ..., 'filling eyebrows',
        'grinding meat', 'knitting'],
       ['headbanging', 'air drumming', 'pumping fist', ...,
        'weaving basket', 'bookbinding', 'skiing slalom'],
       ['exercising arm', 'lunge', 'squat', ..., 'paragliding',
        'biking through snow', 'bee keeping'],
       ...,
       ['waxing chest', 'waxing back', 'waxing legs', ...,
        'skiing crosscountry', 'golf chipping', 'golf driving'],
       ['playing flute', 'playing bass guitar', 'playing clarinet', ...,
        'baby waking up', 'golf chipping', 'washing feet'],
       ['extinguishing fire', 'motorcycling', 'walking the dog', ...,
        'assembling computer', 'waxing chest', 'getting a tattoo']],
      dtype='<U39')

In [22]:
scores

array([[3.7998241e-01, 1.9977430e-01, 1.1454046e-01, ..., 4.1070369e-09,
        4.0600794e-09, 5.2552024e-10],
       [4.4746959e-01, 1.3770965e-01, 7.1373403e-02, ..., 3.1191104e-07,
        1.9676860e-07, 9.9010187e-08],
       [8.8977093e-01, 7.2920687e-02, 1.6560765e-02, ..., 5.9105013e-12,
        5.4985903e-12, 1.6751210e-12],
       ...,
       [7.2241563e-01, 1.0739921e-01, 8.9139961e-02, ..., 4.2127171e-10,
        4.0363674e-10, 1.1107539e-10],
       [9.9448818e-01, 2.4768533e-03, 7.9978246e-04, ..., 6.6125674e-11,
        4.2203865e-11, 1.0055722e-11],
       [8.2344991e-01, 4.8134100e-02, 1.8919127e-02, ..., 9.7456706e-08,
        4.7469481e-08, 5.2609685e-09]], dtype=float32)

Let's store our results as a Pandas DataFrame to make it easier to work with. This dataframe keeps only the top five model predictions for each video. The `Video_Id` index refers to the numbering of the vidoes in our visualize section above.

In [23]:
# collect metadata and model results
results = defaultdict(list)

results['YouTube_Id'] = video_sample['youtube_id'].values
results['Ground_Truth'] = video_sample['label'].values

for s, p in zip(scores, predictions):
    results['scores'].append(list(s[:5]))
    results['preds'].append(list(p[:5]))

results_df = store_results_in_dataframe(results)

In [24]:
results_df

YouTube_Id        Ground_Truth              pred_1  \
video_id                                                        
0         -6KNHzXeYKo             parkour             parkour   
1         -Gio3hF5OA0       brushing hair         headbanging   
2         0wZpjStZtUY      exercising arm      exercising arm   
3         0pQACKjlllc         eating cake    eating ice cream   
4         0uswr636GWs       braiding hair       braiding hair   
5         -JuFFz7KyMo            tickling        waxing chest   
6         -SdDjLIkBHc       playing flute       playing flute   
7         -X1Allr_ZcY  extinguishing fire  extinguishing fire   

                       pred_2            pred_3  \
video_id                                          
0                   hopscotch     skateboarding   
1                air drumming      pumping fist   
2                       lunge             squat   
3            eating spaghetti       eating cake   
4                curling hair       fixing hair   
5                 waxing back       waxing legs   
6         playing bass guitar  playing clarinet   
7                motorcycling   walking the dog   

                                    pred_4          pred_5   score_1  \
video_id                                                               
0                          playing cricket        krumping  0.379982   
1                               beatboxing        slapping  0.447470   
2         exercising with an exercise ball    front raises  0.889771   
3                             blowing nose        laughing  0.348714   
4                               dying hair    washing hair  0.859741   
5                                bandaging        tickling  0.722416   
6                          playing trumpet  playing guitar  0.994488   
7                              pushing car         jogging  0.823450   

           score_2   score_3   score_4   score_5  
video_id                                          
0         0.199774  0.114540  0.093412  0.075015  
1         0.137710  0.071373  0.042713  0.028393  
2         0.072921  0.016561  0.004367  0.003058  
3         0.138570  0.119416  0.054775  0.048254  
4         0.133039  0.006605  0.000457  0.000152  
5         0.107399  0.089140  0.036302  0.019120  
6         0.002477  0.000800  0.000674  0.000506  
7         0.048134  0.018919  0.014241  0.010791

## Evaluating the model
So how well did our model do?  It may seem natural to consider the model's accuracy using only it's top prediction for each video and we'll look at that first. However, when working with hundreds of classes, subtlties arise. For example, some classes could be easily confused -- "catching or throwing a softball" and "catching or throwing a baseball" are both classes but it may be difficult for the model to discern the type of ball in a low quality video or if the ball only takes up a small handful of pixels in a wide-shot video.  Additionaly, videos can contain more than one action -- "texting" while "driving a car" (don't do that!) or "hula hooping" while "playing ukulele". The Kinetics 400 dataset only provides a single ground-truth label for each video, rather than an exhaustive list of annotations. For this reason, the authors recommend evaluating model performance on the top-5 accuracy, rather than top-1. 

### Visualize videos again

Let's first examine the model's top-1 accuracy by considering our video visualization. Our visualization helper function can also accept and display the model's top prediction beneath each video. If the model's prediction does not match the ground truth label, the text will display red. 

In [25]:
video_html = make_video_table(results_df['YouTube_Id'], results_df['Ground_Truth'], results_df['pred_1'])
display.HTML(video_html)

While it may seem alarming that many of these texts are red, consider them in light of the discussion above -- how many of these top-1 labels might be a reasonable description of the video clip, even if it doesn't match the ground truth label? Are any of these labels possible points of confusion or are their multiple actions in the scene? 

### Model accuracy on our very small sample

Finally, let's consider the top-5 accuracy. In this case, we score the model as being "correct" if the ground truth label is within the model's top-5 predictions for a given video. 

In [26]:
#This cell also doesn't run until prediction_df is addressed
accuracy_top_1 = compute_accuracy(results_df, num_top_classes = 1)
accuracy_top_5 = compute_accuracy(results_df, num_top_classes = 5)

top-1 accuracy: 62.50%
top-5 accuracy: 87.50%


As we can see, the model's accuracy improves when we consider the top-5 predictions. While this approach doesn't always make sense for every circumstance, due to the nature of the Kinetics dataset and it's annotations, this method is certainly valid in this case. 

## Evaluating over many videos

So far, everything we've done has been to explore the capability of video classification with a small, concrete example. However, in practice, there are several models and many datasets that one might consider when building a real-world video classification application. In that case, one will need to evaluate dfferent models over various datasets in order to gauge which is most appropriate for the application in question. This is a model benchmarking job. To that end, we've created some additional utilities to facilitate model evaluation over a much larger portion of the Kinetics datasets. Included in this AMP is a benchmarking script that can be automated via the CML Jobs abstraction (or by a simple bash script, if that's your style). Here we provide a quick example of the core utilities contained within that script -- namely, the ability to load, pre-process, and batch over a larger portion of videos. 

We break this task into two parts: load and cache videos, and evaluation. The first step is performed by our `KineticsLoader` class, and does essentially the same steps as our `load_videos` function above. Specifically, this will load the raw `mp4` format into numpy arrays and perform essential pre-processing, such as cropping and resizing to I3D specifications. Once loaded, these video examples are cached (saved to disk) in their numpy format so that they can be reused in any downstream evaluation process. 

In [23]:
num_videos = 64
loader.load_and_cache_video_examples(num_videos)

Processed video already exists at /home/cdsw/data/processed/kinetics/400/val/part_0/0tlGOxUQ0Kw_000074_000084.pkl. Use overwrite=True to re-process.
Processed video already exists at /home/cdsw/data/processed/kinetics/400/val/part_0/-05qSkAhM6Y_000205_000215.pkl. Use overwrite=True to re-process.
Processed video already exists at /home/cdsw/data/processed/kinetics/400/val/part_0/-LISB_b8rIw_000049_000059.pkl. Use overwrite=True to re-process.
Processed video already exists at /home/cdsw/data/processed/kinetics/400/val/part_0/--ILYNHl3e4_000541_000551.pkl. Use overwrite=True to re-process.
Processed video already exists at /home/cdsw/data/processed/kinetics/400/val/part_0/-yv8c2CDbR8_000004_000014.pkl. Use overwrite=True to re-process.
Processed video already exists at /home/cdsw/data/processed/kinetics/400/val/part_0/-aeOuOI3eN0_000219_000229.pkl. Use overwrite=True to re-process.
Processed video already exists at /home/cdsw/data/processed/kinetics/400/val/part_0/-WZgMWx8Elk_000013_000

The second step is an evaluation function that encapsulates all of the prediction steps we performed above. This function accepts a model and data loader class and infers on the requested number of videos, grouping them into the given `batch_size` after resampling to `num_frames` number of frames for each video. The results are stored in a Pandas DataFrame so that we can consider the overall model performance. 

In [25]:
results_df = evaluate(
    i3d400,
    loader, 
    num_videos=num_videos, 
    batch_size=8, 
    top_n_results=5, 
    num_frames=100, 
    savefile="small_test_results_i3d.csv"
)

In [26]:
results_df

YouTube_Id                      Ground_Truth  \
video_id                                                  
0         0mX7eiHEiP8                     setting table   
1         00cwEcZZcu4               biking through snow   
2         -05qSkAhM6Y                     belly dancing   
3         -0r6NmrdKCU                        making tea   
4         018EClOtVTM                             situp   
...               ...                               ...   
59        0d2XdZRy2fc     catching or throwing baseball   
60        -PW6kckornM          punching person (boxing)   
61        0N8szFGAmMw                   carving pumpkin   
62        -9V1qe3Lk2M  exercising with an exercise ball   
63        -5LEMasZdOc                           digging   

                                    pred_1                            pred_2  \
video_id                                                                       
0                               plastering                  dancing macarena   
1                      biking through snow              riding mountain bike   
2                            belly dancing                  contact juggling   
3                               making tea                     setting table   
4                                    situp  exercising with an exercise ball   
...                                    ...                               ...   
59                               high kick               shooting basketball   
60                punching person (boxing)                         wrestling   
61                         carving pumpkin                   balloon blowing   
62        exercising with an exercise ball                       headbutting   
63                           blasting sand                          spraying   

                      pred_3           pred_4           pred_5   score_1  \
video_id                                                                   
0            playing cricket  cleaning toilet           crying  0.098597   
1              riding a bike  riding unicycle   crossing river  0.998980   
2                  hopscotch       plastering          tai chi  0.534826   
3            cleaning toilet    making a cake  shredding paper  0.527350   
4             exercising arm    throwing ball   stretching arm  0.999537   
...                      ...              ...              ...       ...   
59        playing basketball     drop kicking     cheerleading  0.675819   
60               headbutting     drop kicking        side kick  0.991211   
61             making a cake      waxing legs         laughing  0.999981   
62           balloon blowing     pumping fist   drinking shots  0.501306   
63                 hopscotch  watering plants   blowing leaves  0.906689   

           score_2   score_3       score_4       score_5  
video_id                                                  
0         0.065879  0.053389  3.668990e-02  3.604879e-02  
1         0.000945  0.000070  1.514738e-06  1.504889e-06  
2         0.063855  0.042720  2.927776e-02  2.311769e-02  
3         0.123318  0.079832  6.867204e-02  3.120951e-02  
4         0.000294  0.000083  5.420654e-05  1.453274e-05  
...            ...       ...           ...           ...  
59        0.082590  0.074438  3.088997e-02  2.110464e-02  
60        0.003232  0.003036  7.463089e-04  6.455458e-04  
61        0.000011  0.000008  6.156522e-08  4.203142e-08  
62        0.069354  0.061799  5.425986e-02  4.095884e-02  
63        0.033253  0.012355  8.516103e-03  7.819586e-03  

[64 rows x 12 columns]

In [27]:
accuracy_top_1 = compute_accuracy(results_df, num_top_classes = 1)
accuracy_top_5 = compute_accuracy(results_df, num_top_classes = 5)

top-1 accuracy: 56.25%
top-5 accuracy: 75.00%


**If this documentation includes code, including but not limited to, code examples, Cloudera makes this available to you under the terms of the Apache License, Version 2.0, including any required notices. A copy of the Apache License Version 2.0 can be found here.**